### PR TITLE
updating to new datastore/blockstore code with contexts

### PR DESCRIPTION
**Pull request by whyrusleeping**
### Status

The following deps need to be tagged, as we have go mod versions pointing at unstable things:
- github.com/drand/drand

The following deps also exist in the lotus-soup go.mod:
- github.com/drand/drand

### CODE REVIEW

File: go.mod
Comment (aschmahmann) [inline go.mod:74]: would bump to v0.5.1

**Comment by vyzo** [comment]
i'll work on finishing this as we need it to upgrade go-libp2p.

**Comment by vyzo** [comment]
the go-data-transfer dep is broken:

```
go: github.com/filecoin-project/go-data-transfer@v1.11.7-0.20211119001436-c0dbfa5fae4d requires
	github.com/ipfs/go-graphsync@v0.10.6-0.20211119000532-c416dad3bd56: invalid version: unknown revision c416dad3bd56
```

**Comment by vyzo** [comment]
@aarshkshah1992 @dirkmc is the go-data-transfer stuff done?

**Comment by jennijuju** [comment]
this is getting blocked by
 https://github.com/filecoin-project/go-data-transfer/pull/283

which is blocked by 
 https://github.com/filecoin-project/go-ds-versioning/pull/3 

which is blocked by a release tag of  https://github.com/filecoin-project/go-ds-versioning/pull/3

**Comment by arajasek** [comment]
@vyzo go-data-transfer thing should be fixed now

**Comment by vyzo** [comment]
broken go-ipfs-blockstore:
```
go: github.com/ipfs/go-ipfs-blockstore@v1.1.0 requires
	github.com/ipfs/go-datastore@v0.4.7-0.20211013204805-28a3721c2e66: invalid version: unknown revision 28a3721c2e66
```

**Comment by vyzo** [comment]
v1.1.1 also broken:
```
go get github.com/ipfs/go-ipfs-blockstore@v1.1.1
go: github.com/ipfs/go-ipfs-blockstore@v1.1.0 requires
	github.com/ipfs/go-datastore@v0.4.7-0.20211013204805-28a3721c2e66: invalid version: unknown revision 28a3721c2e66
```

**Comment by vyzo** [comment]
This occurs when trying to upgrade go-data-transfer:

```
$ go get github.com/filecoin-project/go-data-transfer@master

go: github.com/filecoin-project/go-data-transfer@v1.11.7-0.20211207053937-e06a599f202a requires
	github.com/ipfs/go-ipfs-blockstore@v1.1.0 requires
	github.com/ipfs/go-datastore@v0.4.7-0.20211013204805-28a3721c2e66: invalid version: unknown revision 28a3721c2e66
```

**Comment by vyzo** [comment]
seems it was a transient go mod cache error of some sorts.

**Comment by vyzo** [comment]
rebased and fixed deps; now the 1M places to add contexts...

**Comment by vyzo** [comment]
Blocker in go-fil-markets:
```
# github.com/filecoin-project/go-fil-markets/shared
../../../../pkg/mod/github.com/filecoin-project/go-fil-markets@v1.13.4/shared/movekey.go:9:20: not enough arguments in call to ds.Has
	have (datastore.Key)
	want (context.Context, datastore.Key)
```

**Comment by vyzo** [comment]
Blocker in go-car:
```
# github.com/ipld/go-car/v2/blockstore
../../../../pkg/mod/github.com/ipld/go-car/v2@v2.1.0/blockstore/readonly.go:24:5: cannot use (*ReadOnly)(nil) (type *ReadOnly) as type blockstore.Blockstore in assignment:
	*ReadOnly does not implement blockstore.Blockstore (wrong type for DeleteBlock method)
		have DeleteBlock(cid.Cid) error
		want DeleteBlock(context.Context, cid.Cid) error
```

**Comment by vyzo** [comment]
blocker in go-storedcounter:
```
# github.com/filecoin-project/go-storedcounter
../../../../pkg/mod/github.com/filecoin-project/go-storedcounter@v0.0.0-20200421200003-1c99c62e8a5b/storedcounter.go:28:23: not enough arguments in call to sc.ds.Has
	have (datastore.Key)
	want (context.Context, datastore.Key)
```

**Comment by vyzo** [comment]
blocker in drand:
```
# github.com/drand/drand/lp2p
../../../../pkg/mod/github.com/drand/drand@v1.2.1/lp2p/ctor.go:92:22: too many arguments in call to libp2p.New
	have (context.Context, ...config.Option)
	want (...config.Option)
```

**Comment by arajasek** [comment]
go-ds-versioning should be unblocked

**Comment by arajasek** [comment]
go-storedcounter should be unblocked

**Comment by arajasek** [comment]
@vyzo started a markets pr here: https://github.com/filecoin-project/go-fil-markets/pull/653

why's go-car pr that maybe works https://github.com/ipld/go-car/pull/273
why's go-ds-versioning pr that maybe works https://github.com/filecoin-project/go-ds-versioning/pull/3

**Comment by vyzo** [comment]
@arajasek can we refrain from force pushing from here on? I have working tree changes that have not been pushed yet.

**Comment by magik6k** [comment]
(`git pull --rebase` should automagically solve all force-push induced issues)

**Comment by arajasek** [comment]
> @arajasek can we refrain from force pushing from here on? I have working tree changes that have not been pushed yet.

Yessir!

**Comment by arajasek** [comment]
@vyzo Okay the only thing this needs should be drand (and the ipldcbor PR linked in description).

**Comment by willscott** [comment]
the drand dep is a bit out of date (v1.2.1 vs current is v1.2.7). I don't see any functional changes to the client in that range though.

PR will be against the next release that we're putting together right now.

**Comment by vyzo** [comment]
All the deps have context now, but a lot of stuff is pointing to non canonical commits; we'll have to make a pass and make sure all the deps are tagged before merging this.

We also want to include go-libp2p v0.17, whose release is imminent.

Other than that, going forward to plumb contexts in all our call sites.

**Comment by vyzo** [comment]
plumbed contexts in the remaining places, `make` succeeds now!

**Comment by vyzo** [comment]
lotus-shed builds now

**Comment by vyzo** [comment]
lotus-bench builds.

**Comment by vyzo** [comment]
everything builds now.

On to fix the tests and lotus-soup.

**Comment by vyzo** [comment]
updated go-libp2p to v0.17 and go-libp2p-connmgr to v0.3; needed (yet another) drand update.

**Comment by vyzo** [comment]
we have DI issues with genesis... something needs a context somewhere and go see Figure 1.

**Comment by vyzo** [comment]
This is getting really messy, we are going to have to squash it.

**Comment by vyzo** [comment]
conformance tests are broken because actors/v2 don't have the context datastores.

This is a blocker, as we don't want to disable those tests. But what can we do? Maybe make a patch release in actors/v2 that has the context datastores?

**Comment by vyzo** [comment]
Status update: everything is ready code-wise, modulo the blocker from actors/v2 and the tagged deps.

**Comment by arajasek** [comment]
Actors v2 now using tagged release

**Comment by arajasek** [comment]
This is now ready to be reviewed and merged.

File: blockstore/badger/blockstore_test.go
Comment (magik6k) [inline blockstore/badger/blockstore_test.go:102]: This is a test, so Background is fine
```suggestion
	ctx := context.Background()
```

File: blockstore/badger/blockstore_test_suite.go
Comment (magik6k) [inline blockstore/badger/blockstore_test_suite.go:47]: Also use .Background here

File: chain/store/messages.go
Comment (magik6k) [inline chain/store/messages.go:32]: Probably worth plumbing context into here / functions below

File: chain/store/snapshot.go
Comment (magik6k) [inline chain/store/snapshot.go:47]: Should be easy to get ctx into here

File: cmd/lotus-bench/import.go
Comment (magik6k) [inline cmd/lotus-bench/import.go:330]: `cctx.Context`?

File: cmd/lotus-miner/init.go
Comment (magik6k) [inline cmd/lotus-miner/init.go:350]: We have `ctx` here (and below)

File: cmd/lotus-miner/init_restore.go
Comment (magik6k) [inline cmd/lotus-miner/init_restore.go:258]: ctx

File: cmd/lotus-sim/simulation/simulation.go
Comment (magik6k) [inline cmd/lotus-sim/simulation/simulation.go:93]: Use .TODO if we don't want to plumb those now

File: cmd/lotus/backup.go
Comment (magik6k) [inline cmd/lotus/backup.go:114]: `cctx.Context`

File: cmd/lotus/backup.go
Comment (magik6k) [inline cmd/lotus/backup.go:117]: cctx.Context

File: cmd/lotus/daemon.go
Comment (magik6k) [inline cmd/lotus/daemon.go:518]: We have ctx here

File: conformance/runner.go
Comment (magik6k) [inline conformance/runner.go:285]:
```suggestion
			if has, err := bs.Has(context.TODO(), link.Cid); err != nil {
```

File: conformance/runner.go
Comment (magik6k) [inline conformance/runner.go:320]:
```suggestion
	_, err = car.LoadCar(context.TODO(), bs, r)
```

File: node/modules/chain.go
Comment (magik6k) [inline node/modules/chain.go:62]: Use `helpers.LifecycleCtx(lc fx.Lifecycle, mctx helpers.MetricsCtx)` to get a 'real' context (or at least cast from mctx helpers.MetricsCtx for metrics)

File: node/modules/chain.go
Comment (magik6k) [inline node/modules/chain.go:86]: `helpers.LifecycleCtx(mctx, lc)`

File: node/modules/client.go
Comment (magik6k) [inline node/modules/client.go:55]: `helpers.LifecycleCtx(mctx, lc)`

File: node/modules/client.go
Comment (magik6k) [inline node/modules/client.go:76]: `helpers.LifecycleCtx(mctx, lc)`

File: node/modules/genesis.go
Comment (magik6k) [inline node/modules/genesis.go:26]: Also use `helpers.LifecycleCtx(mctx, lc)` in this file

File: node/modules/services.go
Comment (magik6k) [inline node/modules/services.go:232]: `helpers.LifecycleCtx(mctx, lc)`

File: node/modules/storageminer.go
Comment (magik6k) [inline node/modules/storageminer.go:81]: Plumb or use TODO

File: node/modules/storageminer.go
Comment (magik6k) [inline node/modules/storageminer.go:303]: `helpers.LifecycleCtx(mctx, lc)`

Comment (magik6k) [reply node/modules/storageminer.go:303]: Same below

File: node/repo/imports/manager.go
Comment (magik6k) [inline node/repo/imports/manager.go:111]: Probably open a tracking issue about this (in markets?)

**Review by magik6k — COMMENTED** [review]
First pass, shouldn't be hard to land.

Was the ffi update intentional?

**Comment by vyzo** [comment]
probably accident, that was not me.

File: chain/store/messages.go
Comment (vyzo) [reply chain/store/messages.go:32]: this will bubble up everywhere, but sure.

File: node/repo/imports/manager.go
Comment (vyzo) [reply node/repo/imports/manager.go:111]: Not sure what it should say and where it should be; what's the problem here?

Comment (magik6k) [reply node/repo/imports/manager.go:111]: IIRC This `Manager` struct implements some interface in markets, it would be nice to have contexts plumbed, but I don't think it needs to happen right now - so we could just open an issue saying 'add contexts to this stuff'

**Review by magik6k — COMMENTED** [review]
Just one nitpick.

Also skimming through the dependency diff quickly (it's.. 30k lines)

**Comment by magik6k** [comment]
(also need to merge master into here to sort out conflicts)

File: blockstore/badger/blockstore_test_suite.go
Comment (vyzo) [reply blockstore/badger/blockstore_test_suite.go:60]: fixed.

File: node/repo/imports/manager.go
Comment (vyzo) [reply node/repo/imports/manager.go:111]: opened #7813.

**Review by magik6k — APPROVED** [review]

**Pull request merged**
